### ISSUE TITLE
bio_dgram_test fails if IPV6 disabled/absent  

### ISSUE THREAD

**Issue by frankozland** [openssl/openssl#25561]
Using https://github.com/openssl/openssl/releases/download/openssl-3.3.2/openssl-3.3.2.tar.gz
Building on raspberry pi 4 - ipv6 is disabled
(same issue as https://github.com/openssl/openssl/issues/19142)

```
04-test_bio_dgram.t ..................... 
        # INFO:  @ test/bio_dgram_test.c:120
        # # Testing with AF_INET, local=0
        # 
        # ERROR: (memory) 'pa == pb' failed @ test/bio_dgram_test.c:55
        # --- pa
        # +++ pb
        # 0000:-c0a80132
        # 0000:+7f000001
        #       ^^^^^^^^
        # 
        # ERROR: (int) 'compare_addr(addr3, addr1) == 1' failed @ test/bio_dgram_test.c:223
        # [0] compared to [1]
        # OPENSSL_TEST_RAND_SEED=1727509514
        not ok 1 - iteration 1
# ------------------------------------------------------------------------------
        # INFO:  @ test/bio_dgram_test.c:120
        # # Testing with AF_INET, local=1
        # 
        # ERROR: (memory) 'pa == pb' failed @ test/bio_dgram_test.c:55
        # --- pa
        # +++ pb
        # 0000:-c0a80132
        # 0000:+7f000001
        #       ^^^^^^^^
        # 
        # ERROR: (int) 'compare_addr(addr3, addr1) == 1' failed @ test/bio_dgram_test.c:223
        # [0] compared to [1]
        # OPENSSL_TEST_RAND_SEED=1727509514
        not ok 3 - iteration 3
# ------------------------------------------------------------------------------
    # OPENSSL_TEST_RAND_SEED=1727509514
    not ok 1 - test_bio_dgram
# ------------------------------------------------------------------------------
../../util/wrap.pl ../../test/bio_dgram_test => 1
not ok 1 - running bio_dgram_test
04-test_bio_dgram.t ..................... 1/? ----------------------------------
#   Failed test 'running bio_dgram_test'
#   at /usr/local/src/openssl-3.3.2/util/perl/OpenSSL/Test/Simple.pm line 77.
04-test_bio_dgram.t ..................... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
04-test_bio_tfo.t ....................... skipped: This test requires enable-tfo
```

**Comment by nhorman** [comment]
I don't think its quite the same.

I assume when you say ipv6 is disabled, you're implying that you've done so dynamically, correct?  i.e. your system config is something include something along the lines of including disable_ipv6=0 on the kernel command line or:

```
echo 1 > /proc/sys/net/ipv6/conf/all/disable_ipv6
```
somewhere?

If so, then the dgram test will fail.  It unilaterally, in bio_dgram_test.c will attempt to use ipv6 addresses if OPENSSL_USE_IPV6 is defined (which is the default).  As a result, any syscall that attempts to create an ipv6 socket will fail, resulting in the subsequent failure you see above.

Whats needed here is something similar to what was done in the 80-test_cmp_http.t test using the have_IPv6() function from Utill.pm

In summary, whats needed is:
1) Some additional code in test/bio_dgram_test.c to add either a command line switch or environment variable to skip ipv6 testing
2) An augmentation to the test/reciepies/04-test_bio_dgram.t file to call have_IPv6() and set the environment variable/command line switch from (1) if ipv6 isn't available

The idea being to make the bio_dgram_test sensitive to ipv6 availability at run time rather than just at compile time.

Is this something you feel like you could create a PR for?

**Comment by nhorman** [comment]
this was fixed by https://github.com/openssl/openssl/pull/24748
